ecosystem/typescript/sdk_v2/src/internal/transaction.ts

```diff
@@ -32,3 +34,142 @@ export async function getGasPriceEstimation(args: { aptosConfig: AptosConfig })
   });
   return data;
 }
+
+export async function getTransactionByVersion(args: {
+  aptosConfig: AptosConfig;
+  txnVersion: AnyNumber;
+}): Promise<TransactionResponse> {
+  const { aptosConfig, txnVersion } = args;
+  const { data } = await getAptosFullNode<{}, TransactionResponse>({
+    aptosConfig,
+    originMethod: "getTransactionByVersion",
+    path: `transactions/by_version/${txnVersion}`,
+  });
+  return data;
+}
+
+export async function getTransactionByHash(args: {
+  aptosConfig: AptosConfig;
+  txnHash: HexInput;
+}): Promise<TransactionResponse> {
+  const { aptosConfig, txnHash } = args;
+  const { data } = await getAptosFullNode<{}, TransactionResponse>({
+    aptosConfig,
+    path: `transactions/by_hash/${txnHash}`,
+    originMethod: "getTransactionByHash",
+  });
+  return data;
+}
+
+export async function isTransactionPending(args: { aptosConfig: AptosConfig; txnHash: HexInput }): Promise<boolean> {
+  const { aptosConfig, txnHash } = args;
+  try {
+    const transaction = await getTransactionByHash({ aptosConfig, txnHash });
+    return transaction.type === TransactionResponseType.Pending;
+  } catch (e: any) {
+    if (e?.status === 404) {
+      return true;
+    }
+    throw e;
+  }
+}
+
+export async function waitForTransaction(args: {
+  aptosConfig: AptosConfig;
+  txnHash: HexInput;
+  extraArgs?: { timeoutSecs?: number; checkSuccess?: boolean };
+}): Promise<TransactionResponse> {
+  const { aptosConfig, txnHash, extraArgs } = args;
+  const timeoutSecs = extraArgs?.timeoutSecs ?? DEFAULT_TXN_TIMEOUT_SEC;
+  const checkSuccess = extraArgs?.checkSuccess ?? true;
+
+  let isPending = true;
+  let count = 0;
+  let lastTxn: TransactionResponse | undefined;
+  let lastError: AptosApiError | undefined;
+
+  while (isPending) {
+    if (count >= timeoutSecs) {
+      break;
+    }
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      lastTxn = await getTransactionByHash({ aptosConfig, txnHash });
+
+      isPending = lastTxn.type === TransactionResponseType.Pending;
+
+      if (!isPending) {
+        break;
+      }
+    } catch (e) {
+      // In short, this means we will retry if it was an AptosApiError and the code was 404 or 5xx.
+      const isAptosApiError = e instanceof AptosApiError;
+      if (!isAptosApiError) {
+        throw e; // This would be unexpected
+      }
+      lastError = e;
+      const isRequestError = e.status !== 404 && e.status >= 400 && e.status < 500;
+      if (isRequestError) {
+        throw e;
+      }
+    }
+    // eslint-disable-next-line no-await-in-loop
+    await sleep(1000);
```

Comment:
Seems long, so minimum a transaction will take 1 second regardless of whether it's shorter.

May want to look into an exponential backoff or something like that.